### PR TITLE
use asyncio.run to run gen_cluster, gen_test and cluster

### PR DESCRIPTION
**Pull request by graingert**
- [x] Closes https://github.com/dask/distributed/issues/6164
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Review by fjetter — COMMENTED** [review]
CI is very sad

File: distributed/utils_test.py [hunk 1]
```diff
@@ -1010,161 +1053,156 @@ def _(func):
...
-                                    args = [s] + workers
-                                    break
-                            if s is False:
-                                raise Exception("Could not start cluster")
-                            if client:
-                                c = await Client(
-                                    s.address,
-                                    loop=loop,
+            async def async_fn():
+                result = None
```
[the remark this answers is not in the document]
Comment (graingert) [reply distributed/utils_test.py:1059]: there's actually a few changes, eg I don't pass `loop` to `await start_cluster(` and `await Client(`

**Review by fjetter — APPROVED** [review]
I'm OK with the changes proposed. If test failures turn out to be unrelated I'm good to go. 

I left a few review comments where I think more inline commentary would help but this is not a requirement.

**Comment by graingert** [comment]
looks like the test suite ending with exit code 1 and all tests passing is already happening: https://github.com/dask/distributed/runs/6604868888?check_suite_focus=true#step:11:1852 (from https://github.com/dask/distributed/issues/6451)

**Comment by graingert** [comment]
ubuntu-latest 3.8 c1: https://github.com/dask/distributed/issues/6452
ubuntu-latest 3.10 not ci1: https://github.com/dask/distributed/issues/6453 https://github.com/dask/distributed/issues/6305
windows-latest 3.10 not c1: https://github.com/dask/distributed/issues/6454
gpuCI: https://github.com/dask/distributed/issues/6429

File: distributed/utils_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (graingert) [reply distributed/utils_test.py:1068]: yep it's just the avoiding passing a loop kwarg, moving clean to a decorator, and using `asyncio.run` with `asyncio.wait_for` instead of run_sync

**Comment by mrocklin** [comment]
> It's introducing stricter cleanup and cancelling tasks instead of allowing them to be garbage collected, so it's likely to introduce more flaky tests.

I'm curious about this.  Is there some useful signal that we can give to future developers about why a test is flaky in this way that would help them?  I ask this question without having thought deeply about it.

**Comment by mrocklin** [comment]
Or, more broadly, do you have thoughts on how to address the increase in flaky tests that we're likely to see?

I like that we're flushing these issues out into the open, but I'd also like to get some ideas on how we can address them.  Is there any systemic approach that you see?

**Comment by graingert** [comment]
> > It's introducing stricter cleanup and cancelling tasks instead of allowing them to be garbage collected, so it's likely to introduce more flaky tests.
> 
> I'm curious about this. Is there some useful signal that we can give to future developers about why a test is flaky in this way that would help them? I ask this question without having thought deeply about it.

This is mostly about edge cases involving implicitly closed rpc objects and implicitly closed tasks.

Implicitly closed tasks are now explicitly cancelled and so await calls will start raising CancelledError during the test rather than GeneratorExit when the task is garbage collected.
